llvm/lib/Target/X86/X86NoginCountInstsPass.cpp

```diff
@@ -0,0 +1,50 @@
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86Subtarget.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+
+using namespace llvm;
+
+class X86NoginCountInstsPass : public MachineFunctionPass {
+public:
+  static inline char ID = 0;
+
+  X86NoginCountInstsPass() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MachFunc) override {
+    DebugLoc DebugLocation = MachFunc.front().begin()->getDebugLoc();
+    const TargetInstrInfo *InstrInfo = MachFunc.getSubtarget().getInstrInfo();
+    size_t CountInstsReg =
+        MachFunc.getRegInfo().createVirtualRegister(&X86::GR64RegClass);
+    size_t Count = 0;
+
+    for (auto &BasicBlock : MachFunc) {
+      Count += countInstrBasBlock(BasicBlock);
+    }
+
+    BuildMI(MachFunc.back(), MachFunc.back().getFirstTerminator(),
```

Comment:
Last block in machine functions blocks list is not necessarily the last one in real function containing return instruction.
Moreover, there may be several basic blocks with different instructions terminating function call.

Reply:
Still there can be more than one block with termination instruction (and it is not only return) in a function.